### ISSUE TITLE
"CS" in the mod overlay isn't changed accordingly to the mode chosen.

### ISSUE THREAD

**Issue by UnknownRedFoxDev** [ppy/osu#25730]
### Type

Cosmetic

### Bug description

When opening the mod context menu, you can see different informations concerning the map's settings, such as circle size, approche rate, etc.. However, when changing mode the "CS" text doesn't change accordingly.   
Meaning that, for standard mode it's fine, but when switching for example to mania, the text is still "CS", when it would be "KC" short for Key Count, as the value below is displayed correctly, showing correctly the key count.

### Screenshots or videos

![image](https://github.com/ppy/osu/assets/114443299/b8368f68-ba4f-4817-828c-56df3b9950e5)

### Version

Lazer 2023.1130.0

### Logs

[performance.log](https://github.com/ppy/osu/files/13641403/performance.log)
[runtime.log](https://github.com/ppy/osu/files/13641404/runtime.log)
[updater.log](https://github.com/ppy/osu/files/13641405/updater.log)
[database.log](https://github.com/ppy/osu/files/13641406/database.log)
[input.log](https://github.com/ppy/osu/files/13641407/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/13641408/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/13641409/network.log)

**Comment by toadsworth-jr** [comment]
Maybe a duplicate of https://github.com/ppy/osu/issues/24978 ?

**Comment by Joehuu** [comment]
Not a duplicate. Needs two different changes.

**Comment by ReflexLevel0** [comment]
It turns out that it's not just the issue with the mod screen label.
![osu_2023-12-16_18-07-22](https://github.com/ppy/osu/assets/47285986/86a92d77-f39c-4c54-ac8b-4e4fe378582a)
![osu_2023-12-16_18-19-58](https://github.com/ppy/osu/assets/47285986/09baf613-8499-47d4-8085-510987daa8a5)
As can be seen from the screenshots, it turns out that in song select value for Key Count is 4, while in mod manager it shows a value of 2.5 for CS, so it seems to be two different values.

**Comment by bdach** [comment]
@ReflexLevel0 that's because you're viewing a converted beatmap on that screenshot there. it's somewhat related.

when we're on the topic the thing shouldn't have a decimal point in mania (what is a decimal key after all?). and key mods should change its value.

but this can be generalized even further because both taiko and catch display attributes in that same display that do nothing for those rulesets. so ideally the way that this would be fixed is by generalizing the entire structure to remove hardcoding of attributes and their presence or lack thereof.

**Comment by ReflexLevel0** [comment]
Yeah we could definitely look into not doing hardcoding, but first I have to figure out how to get Key Count data.
Currently, all of the data about beatmap attributes is pulling data from BeatmapDifficulty class, as can be seen in the following code: https://github.com/ppy/osu/blob/1ebb6262fd9bb01d82f135bccc7fff11a31abbe6/osu.Game/Overlays/Mods/BeatmapAttributesDisplay.cs#L186-L189
However that class does not contain any data about KeyCount, it only has DrainRate, CircleSize, OverallDifficulty, ApproachRate, SliderMultiplier and SliderTickRate.
Does anyone have any idea how to get the KeyCount value?

**Comment by ReflexLevel0** [comment]
@bdach Also when you say "so ideally the way that this would be fixed is by generalizing the entire structure to remove hardcoding of attributes and their presence or lack thereof", which attributes exactly should be excluded from each?
I'm guessing you are thinking about updating file https://github.com/ppy/osu/blob/1ebb6262fd9bb01d82f135bccc7fff11a31abbe6/osu.Game/Beatmaps/IBeatmapDifficultyInfo.cs so that it contains only data relevant to all modes, like taiko not having drain rate, and then have different rulesets inherit from that interface and add their own attributes, or am I missing the mark?
Also doing that might be outside of the scope of this issue so it might be better to first solve the issue with the key count and then create a new issue to do this.

**Comment by bdach** [comment]
> which attributes exactly should be excluded from each?

check what attributes difficulty adjust supports in each game mode.

> I'm guessing you are thinking about updating file https://github.com/ppy/osu/blob/1ebb6262fd9bb01d82f135bccc7fff11a31abbe6/osu.Game/Beatmaps/IBeatmapDifficultyInfo.cs so that it contains only data relevant to all modes, like taiko not having drain rate, and then have different rulesets inherit from that interface and add their own attributes, or am I missing the mark?

yes, you are missing the mark, you wouldn't touch that structure, that is way too much work for this one issue. you'd fix it locally to the display.

**Comment by ReflexLevel0** [comment]
ohhh okay I think I get it, I just have to find a way to get Key Count and then I can fix the issue, thanks for the help :)

**Comment by bdach** [comment]
loose sketch of the solution would be to make the display ask the ruleset what should be shown and act as a dumb display from that point on (logic-wise).

**Comment by ReflexLevel0** [comment]
So my idea of how this might look can be seen in the video:
https://github.com/ppy/osu/assets/47285986/3f2bac0e-f7f0-4565-979a-4c523a9dc9af
It shows or hides attributes based on the current ruleset (right now I have hard coded this so that what attributes get shown are based on current ruleset, but if there is a way to get that data dynamically from the current ruleset I could do that too).

```csharp
var rightContentDisplays = new List<VerticalAttributeDisplay>();
switch (gameRuleset.Value.Name)
{
    case "osu!":
    case "osu!catch":
        rightContentDisplays.AddRange(new[] { circleSizeDisplay, drainRateDisplay, overallDifficultyDisplay, approachRateDisplay });
        break;
    case "osu!taiko":
        rightContentDisplays.AddRange(new[] { overallDifficultyDisplay });
        break;
    case "osu!mania":
        rightContentDisplays.AddRange(new[] { keyCountDisplay, drainRateDisplay, overallDifficultyDisplay });
        break;
    default:
        Console.WriteLine(gameRuleset.Value.Name);
        throw new NotImplementedException("Mod attribute screen only supports standard, taiko, catch and mania rulesets right now");
}
RightContent.RemoveAll(_ => true, false);
RightContent.AddRange(rightContentDisplays);
```
This is just temporary code for testing and to see if my idea make sense (for example, when commiting actual code, NotImplementedException shouldn't be thrown just because rulesets are hardcoded, it would be better to log an error message instead).

**Comment by ReflexLevel0** [comment]
Edit: if nothing it's probably better to have 
```csharp
default:
  rightContentDisplays.AddRange(new[] { circleSizeDisplay, drainRateDisplay, overallDifficultyDisplay, approachRateDisplay });
  break;
```
then for it to crash if ruleset isn't standard, taiko, mania or catch the beat.

**Comment by ReflexLevel0** [comment]
> @ReflexLevel0 that's because you're viewing a converted beatmap on that screenshot there. it's somewhat related.

Just a question regarding the converted beatmap thing, is there a way to view the original one? Because then I would be able to get the correct key value for mania (from what i know).

**Comment by Joehuu** [comment]
Superseded by https://github.com/ppy/osu/issues/34236.
